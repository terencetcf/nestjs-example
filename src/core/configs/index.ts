export * from './app.config';
export * from './http-server.config';
