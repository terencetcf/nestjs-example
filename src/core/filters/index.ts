export * from './exception.filter';
export * from './global.filter';
export * from './http-exception.filter';
