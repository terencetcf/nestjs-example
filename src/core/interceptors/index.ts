export * from './global.interceptor';
export * from './response.interceptor';
