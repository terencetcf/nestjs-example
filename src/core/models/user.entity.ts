import { Column, Entity, PrimaryColumn } from 'typeorm';

@Entity()
export class User {
  @PrimaryColumn({ type: 'uuid' })
  id!: string;

  @Column({
    length: 50,
  })
  username!: string;

  @Column({
    length: 100,
  })
  first_name!: string;

  @Column({
    length: 100,
  })
  last_name!: string;
}
