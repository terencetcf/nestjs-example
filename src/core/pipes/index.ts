export * from './global.pipe';
export * from './parse-number-id.pipe';
