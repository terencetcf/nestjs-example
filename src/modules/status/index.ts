export * from './dto';
export * from './status.controller';
export * from './status.module';
export * from './status.service';
