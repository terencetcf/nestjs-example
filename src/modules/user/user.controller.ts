import { Body, Controller, Get, HttpStatus, Post } from '@nestjs/common';
import { ApiOperation, ApiResponse } from '@nestjs/swagger';

import { CreateUserDto, UserDto } from './dto';
import { UserService } from './user.service';

@Controller()
export class UserController {
  constructor(private readonly userService: UserService) {}

  @ApiOperation({
    summary: 'Get users',
  })
  @ApiResponse({
    status: HttpStatus.OK,
    type: UserDto,
    isArray: true,
    description: 'List of users',
  })
  @Get('users')
  async getUsers(): Promise<UserDto[]> {
    const result = await this.userService.getUsers();

    return result;
  }

  @ApiOperation({
    summary: 'Create user',
  })
  @ApiResponse({
    status: HttpStatus.CREATED,
    description: 'User created',
    type: UserDto,
  })
  @ApiResponse({
    status: HttpStatus.BAD_REQUEST,
    description: `Invalid payload`,
  })
  @Post('users')
  async createUser(@Body() payload: CreateUserDto): Promise<UserDto> {
    const result = await this.userService.createUser(payload);

    return result;
  }
}
