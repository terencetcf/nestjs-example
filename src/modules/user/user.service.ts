import { Injectable } from '@nestjs/common';
import { InjectRepository } from '@nestjs/typeorm';
import { User } from '@tt/core/models';
import { generatorUtil } from '@tt/utils';
import { Repository } from 'typeorm';

import { CreateUserDto, UserDto } from './dto';

@Injectable()
export class UserService {
  constructor(
    @InjectRepository(User)
    private readonly userRepository: Repository<User>,
  ) {}

  async getUsers(): Promise<UserDto[]> {
    const users = await this.userRepository.find();

    const dto = users.map((user) => new UserDto(user));

    return dto;
  }

  async createUser(createUserDto: CreateUserDto): Promise<UserDto> {
    const mappedUser: Partial<User> = {
      id: generatorUtil.uuid(),
      username: createUserDto.username,
      first_name: createUserDto.first_name,
      last_name: createUserDto.last_name,
    };

    const createdUser = await this.userRepository.save(mappedUser);

    return new UserDto(createdUser);
  }
}
