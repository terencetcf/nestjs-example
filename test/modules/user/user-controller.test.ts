import { HttpStatus } from '@nestjs/common';
import { User } from '@tt/core/models';
import { UserController, UserService } from '@tt/modules/user';
import { TestApp } from '@tt/test/setup';

import { testData } from './__fixtures__';

describe('UserController', () => {
  let app: TestApp;
  beforeAll(async () => {
    app = await TestApp.create().setup(
      [UserController],
      [UserService],
      [User],
      testData.entitiesToBeAddedToDb,
    );
  });

  afterAll(async () => {
    await app.close();
  });

  describe('GET /users', () => {
    it('should return OK status with expected response body', async () => {
      // Act
      const actual = await app.httpRequest.get(`/users`);

      // Assert
      expect(actual.status).toEqual(HttpStatus.OK);
      expect(actual.body).toMatchSnapshot();
    });
  });

  describe('POST /users', () => {
    it('should return OK status with expected response body', async () => {
      // Act
      const actual = await app.httpRequest
        .post(`/users`)
        .send(testData.createUserDto);

      // Assert
      expect(actual.status).toEqual(HttpStatus.CREATED);
      expect(actual.body).toMatchSnapshot({
        data: {
          id: expect.any(String),
        },
      });
    });

    it('should return BadRequest status if payload is invalid', async () => {
      // Arrange
      const invalidPayload = {};

      // Act
      const actual = await app.httpRequest.post(`/users`).send(invalidPayload);

      // Assert
      expect(actual.status).toEqual(HttpStatus.BAD_REQUEST);
      expect(actual.body).toMatchSnapshot();
    });

    it('should return BadRequest status if payload exceeded limits', async () => {
      // Arrange
      const invalidPayload = {
        username: 'x'.repeat(51),
        first_name: 'x'.repeat(101),
        last_name: 'x'.repeat(101),
      };

      // Act
      const actual = await app.httpRequest.post(`/users`).send(invalidPayload);

      // Assert
      expect(actual.status).toEqual(HttpStatus.BAD_REQUEST);
      expect(actual.body).toMatchSnapshot();
    });
  });
});
