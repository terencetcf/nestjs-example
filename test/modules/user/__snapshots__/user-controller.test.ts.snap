// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserController GET /users should return OK status with expected response body 1`] = `
Object {
  "data": Array [
    Object {
      "first_name": "John",
      "id": "070b50ca-486a-41e7-9a2e-cda2c16204ac",
      "last_name": "Doe",
      "username": "test_user",
    },
  ],
}
`;

exports[`UserController POST /users should return BadRequest status if payload exceeded limits 1`] = `
Object {
  "error": Object {
    "message": Array [
      "username must be shorter than or equal to 50 characters",
      "first_name must be shorter than or equal to 100 characters",
      "last_name must be shorter than or equal to 100 characters",
    ],
  },
}
`;

exports[`UserController POST /users should return BadRequest status if payload is invalid 1`] = `
Object {
  "error": Object {
    "message": Array [
      "username must be longer than or equal to 2 characters",
      "username should not be empty",
      "first_name must be longer than or equal to 2 characters",
      "first_name should not be empty",
      "last_name must be longer than or equal to 2 characters",
      "last_name should not be empty",
    ],
  },
}
`;

exports[`UserController POST /users should return OK status with expected response body 1`] = `
Object {
  "data": Object {
    "first_name": "Steve",
    "id": Any<String>,
    "last_name": "Job",
    "username": "new-username",
  },
}
`;
